File: solutions/ts/22.ts
function generateGiftSets(gifts: string[]): string[][] {
  const result: string[][] = [];
  function generateCombinations(
    arr: string[],
    size: number,
    start: number,
    temp: string[]
  ) {
    if (temp.length === size) {
      result.push([...temp]);
      return;
    }
    for (let i = start; i < arr.length; i++) {
      temp.push(arr[i]);
      generateCombinations(arr, size, i + 1, temp);
      temp.pop();
    }
  }
  for (let i = 1; i <= gifts.length; i++) {
    generateCombinations(gifts, i, 0, []);
  }
  return result;
}

const giftsSet1 = generateGiftSets(["car", "doll", "puzzle"]);
console.log(giftsSet1);
// [
//   ['car'],
//   ['doll'],
//   ['puzzle'],
//   ['car', 'doll'],
//   ['car', 'puzzle'],
//   ['doll', 'puzzle'],
//   ['car', 'doll', 'puzzle']
// ]

const giftsSet2 = generateGiftSets(["ball"]);
console.log(giftsSet2);
// [
//   ['ball']
// ]

const giftsSet3 = generateGiftSets(["game", "pc"]);
console.log(giftsSet3);
// [
//   ['game'],
//   ['pc'],
//   ['game', 'pc']
// ]

/* 
  Score: ★★★★★
*/

File: solutions/python/22.py
def generate_gift_sets(gifts):
    result = []

    def generate_combinations(arr, size, start, temp):
        if (len(temp) == size):
            result.append([*temp])
            return
        i = start
        while (i < len(arr)):
            temp.append(arr[i])
            generate_combinations(arr, size, i+1, temp)
            temp.pop()
            i += 1
    for i in range(len(gifts)+1)[1:]:
        generate_combinations(gifts, i, 0, [])
    return result


if __name__ == "__main__":
    gifts_set_1 = generate_gift_sets(["car", "doll", "puzzle"])
    print(gifts_set_1)
    # [
    #   ['car'],
    #   ['doll'],
    #   ['puzzle'],
    #   ['car', 'doll'],
    #   ['car', 'puzzle'],
    #   ['doll', 'puzzle'],
    #   ['car', 'doll', 'puzzle']
    # ]

    gifts_set_2 = generate_gift_sets(["ball"])
    print(gifts_set_2)
    # [
    #   ['ball']
    # ]

    gifts_set_3 = generate_gift_sets(["game", "pc"])
    print(gifts_set_3)
    # [
    #   ['game'],
    #   ['pc'],
    #   ['game', 'pc']
    # ]

"""
  Score: ★★★★★
"""
